Sources/NIOPosix/LinuxUring.swift

```diff
@@ -38,84 +40,135 @@ internal extension TimeAmount {
-        assert(MemoryLayout<UInt64>.size == MemoryLayout<URingUserData>.size)
-        self.registrationID = UInt16(truncatingIfNeeded: registrationID.rawValue)
-        self.fileDescriptor = fileDescriptor
-        self.eventType = eventType
-        self.padding = 0
+// These are the events returned up to the selector
+internal struct URingEvent {
+    enum EventType {
+        case _none
+        case poll(UInt32, Bool) // poll mask, poll cancelled
+        case write(Int32)
     }
 
-    @inlinable init(rawValue: UInt64) {
-        let unpacked = IntegerBitPacking.unpackUInt32UInt16UInt8(rawValue)
-        self = .init(registrationID: SelectorRegistrationID(rawValue: UInt32(unpacked.1)),
-                     fileDescriptor: CInt(unpacked.0),
-                     eventType: CQEEventType(rawValue:unpacked.2)!)
-    }
-}
-
-extension UInt64 {
-    init(_ uringUserData: URingUserData) {
-        let fd = uringUserData.fileDescriptor
-        let eventType = uringUserData.eventType.rawValue
-        assert(fd >= 0, "\(fd) is not a valid file descriptor")
-        assert(eventType >= 0, "\(eventType) is not a valid eventType")
+    var fd: CInt
+    var registrationID: UInt16 // we just have the truncated lower 16 bits of the registrationID
+    var type: EventType
 
-        self = IntegerBitPacking.packUInt32UInt16UInt8(UInt32(truncatingIfNeeded: fd),
-                                                       uringUserData.registrationID,
-                                                       eventType)
+    init() {
+        fd = 0
+        registrationID = 0
+        type = ._none
     }
-}
 
-// These are the events returned up to the selector
-internal struct URingEvent {
-    var fd: CInt
-    var pollMask: UInt32
-    var registrationID: UInt16 // we just have the truncated lower 16 bits of the registrationID
-    var pollCancelled: Bool
-    init () {
-        self.fd = -1
-        self.pollMask = 0
-        self.registrationID = 0
-        self.pollCancelled = false
+    init(_ fd: CInt, _ registrationID: UInt16, _ type: EventType) {
+        self.fd = fd
+        self.registrationID = registrationID
+        self.type = type
     }
 }
 
 // This is the key we use for merging events in our internal hashtable
-struct FDEventKey: Hashable {
-    var fileDescriptor: CInt
-    var registrationID: UInt16 // we just have the truncated lower 16 bits of the registrationID
+private struct EventKey: Hashable {
+    let fileDescriptor: CInt
+    let registrationID: UInt16 // we just have the truncated lower 16 bits of the registrationID
 
-    init(_ f: CInt, _ s: UInt16) {
-        self.fileDescriptor = f
-        self.registrationID = s
+    init(_ fileDescriptor: CInt, _ registrationID: UInt16) {
+        self.fileDescriptor = fileDescriptor
+        self.registrationID = registrationID
     }
 }
 
+private func _debugPrint(_ s: @autoclosure () -> String) {
+    #if SWIFTNIO_IO_URING_DEBUG
```

Comment:
That condition was before I started to work on it, just left it...

Reply:
I think it's a remnant of the original bringup to not having to define out every debugPrint that is in place IIRC.

There are additionally three more specific defines for more in-depth debugging, but they were in few places and are completely defined out (https://github.com/search?q=repo%3Aapple%2Fswift-nio%20SWIFTNIO_IO_URING_DEBUG&type=code) .

Perhaps there's a better way to do this (maybe with the macro support coming up? Basically, one would like to be able to build with debug logging during development but to completely remove traces of it for production as this may be a fairly performance sensitive thing).